src/Lights/Shadows/cascadedShadowGenerator.ts

```diff
@@ -0,0 +1,1708 @@
+
+        if (!scene.shadowsEnabled || !light.shadowEnabled) {
+            return;
+        }
+
+        defines["SHADOW" + lightIndex] = true;
+
+        defines["SHADOWCSM" + lightIndex] = true;
+        defines["SHADOWCSMDEBUG" + lightIndex] = this.debug;
+        defines["SHADOWCSMNUM_CASCADES" + lightIndex] = this.numCascades;
+
+        const camera = scene.activeCamera;
+
+        if (camera && this._shadowMaxZ < camera.maxZ) {
+            defines["SHADOWCSMUSESHADOWMAXZ" + lightIndex] = true;
+        }
+
+        if (this.useContactHardeningShadow) {
+            defines["SHADOWPCSS" + lightIndex] = true;
+            if (this._filteringQuality === CascadedShadowGenerator.QUALITY_LOW) {
+                defines["SHADOWLOWQUALITY" + lightIndex] = true;
+            }
+            else if (this._filteringQuality === CascadedShadowGenerator.QUALITY_MEDIUM) {
+                defines["SHADOWMEDIUMQUALITY" + lightIndex] = true;
+            }
+            // else default to high.
+        }
+        else if (this.usePercentageCloserFiltering) {
+            defines["SHADOWPCF" + lightIndex] = true;
+            if (this._filteringQuality === CascadedShadowGenerator.QUALITY_LOW) {
+                defines["SHADOWLOWQUALITY" + lightIndex] = true;
+            }
+            else if (this._filteringQuality === CascadedShadowGenerator.QUALITY_MEDIUM) {
+                defines["SHADOWMEDIUMQUALITY" + lightIndex] = true;
+            }
+            // else default to high.
+        }
+    }
+
+    /**
+     * Binds the shadow related information inside of an effect (information like near, far, darkness...
+     * defined in the generator but impacting the effect).
+     * @param lightIndex Index of the light in the enabled light list of the material owning the effect
+     * @param effect The effect we are binfing the information for
+     */
+    public bindShadowLight(lightIndex: string, effect: Effect): void {
+        const light = this._light;
+        const scene = this._scene;
+
+        if (!scene.shadowsEnabled || !light.shadowEnabled) {
+            return;
+        }
+
+        const camera = scene.activeCamera;
+        if (!camera) {
+            return;
+        }
+
+        const shadowMap = this.getShadowMap();
+        if (!shadowMap) {
+            return;
+        }
+
+        const width = shadowMap.getSize().width;
+
+        for (let cascadeIndex = 0; cascadeIndex < this._numCascades; ++cascadeIndex) {
+            this._lightSizeUVCorrection[cascadeIndex * 2 + 0] = cascadeIndex === 0 ? 1 : (this._cascadeMaxExtents[0].x - this._cascadeMinExtents[0].x) / (this._cascadeMaxExtents[cascadeIndex].x - this._cascadeMinExtents[cascadeIndex].x); // x correction
+            this._lightSizeUVCorrection[cascadeIndex * 2 + 1] = cascadeIndex === 0 ? 1 : (this._cascadeMaxExtents[0].y - this._cascadeMinExtents[0].y) / (this._cascadeMaxExtents[cascadeIndex].y - this._cascadeMinExtents[cascadeIndex].y); // y correction

```

Comment:
Peut on cacher ces valeurs au compute des extents ?

Reply:
Pas vraiment parce qu'on a besoin de la cascade 0 en plus de la cascade courante pour faire les calculs. Or quand on est dans `getCSMTransformation(cascadeIndex)` (avec cascadeIndex > 0) qui est l'endroit où on calcule les extents, rien ne dit que getCSMTransformation(0) aura été appelée avant car l'appel est fait dans un `if (this.isReady(subMesh, ...)`. Donc, si pour une raison ou une autre on n'est jamais passé là pour la cascade 0 mais qu'on y passe pour la 1, ça peut merder...

Par contre, un truc que je pourrais faire c'est de faire tous les calculs des X cascades dans RTT.onBeforeBindObservable() et pas dans getCSMTransformation(cascadeIndex) (qui se contenterait alors de juste retourner la matrice déjà calculée): de toute façon on a besoin de faire les calculs pour toutes les cascades à chaque frame, pas besoin d'attendre que getCSMTransformation() soit appelée pour ça.

A noter que dans tous les cas il n'y aurait pas de caching: ce calcul est effectué à chaque frame car l'extent des cascades peut varier d'une frame à l'autre.

Un truc que je vais faire de toute façon c'est de déplacer ces lignes dans le `if (PCSS)`, vu que ça ne sert que dans ce cas.